firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java

```diff
@@ -29,32 +30,51 @@
 
   private final Provider<HeartBeatInfo> heartBeatInfoProvider;
   private final Provider<UserAgentPublisher> userAgentPublisherProvider;
-  private final String firebaseFirestoreHeartBeatTag = "fire-fst";
+  private final FirebaseOptions firebaseOptions;
+
+  private static final String HEART_BEAT_TAG = "fire-fst";
 
   private static final Metadata.Key<String> HEART_BEAT_HEADER =
       Metadata.Key.of("x-firebase-client-log-type", Metadata.ASCII_STRING_MARSHALLER);
 
   private static final Metadata.Key<String> USER_AGENT_HEADER =
       Metadata.Key.of("x-firebase-client", Metadata.ASCII_STRING_MARSHALLER);
 
+  private static final Metadata.Key<String> GMP_APP_ID_HEADER =
+      Metadata.Key.of("x-firebase-gmpid", Metadata.ASCII_STRING_MARSHALLER);
+
   public FirebaseClientGrpcMetadataProvider(
       @NonNull Provider<UserAgentPublisher> userAgentPublisherProvider,
-      @NonNull Provider<HeartBeatInfo> heartBeatInfoProvider) {
+      @NonNull Provider<HeartBeatInfo> heartBeatInfoProvider,
+      FirebaseOptions firebaseOptions) {
```

Comment:
Declare this `@NonNull` or `@Nullable`.

Reply:
Done (went with `@Nullable` mostly because I'm not really sure how this thing with dependency injection works and accounting for null seemed safer).

Reply:
I suggest you use `@NonNull` here as `FirebaseOptions` are always present, I would also (optionally) recommend changing your dependency on `FirebaseOptions` to `required` here: https://github.com/firebase/firebase-android-sdk/blob/22d1615474da1bfb71652447f65c0797818ecdb7/firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java#L51